core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/interceptor/RequestCompressionInterceptor.java

```diff
@@ -85,8 +86,19 @@ public Optional<RequestBody> modifyHttpContent(Context.ModifyHttpRequest context
         Compressor compressor = resolveCompressionType(executionAttributes);
         RequestBody requestBody = context.requestBody().get();
 
-        // TODO - Sync streaming compression implementation
-        throw new UnsupportedOperationException();
+        if (isTransferEncodingChunked(context)) {
+            InputStream compressedStream = compressor.compress(requestBody.contentStreamProvider().newStream());
+            try {
+                byte[] compressedBytes = IoUtils.toByteArray(compressedStream);
+                return Optional.of(RequestBody.fromBytes(compressedBytes));
```

Comment:
Why do we want to load the whole payload into memory when we are using chunked encoding?

Reply:
if `TransferEncoding : chunked` header is present, we want to compress the entire payload before it is chunked (per specification)